libraries/AC_WPNav/AC_WPNav.cpp

```diff
@@ -202,93 +242,79 @@ bool AC_WPNav::get_wp_destination(Location& destination) const
     }
 
-    // set origin and destination
-    return set_wp_origin_and_destination(origin, destination, terrain_alt);
-}
+    // update destination
+    _destination = destination;
+    _terrain_alt = terrain_alt;
 
-/// set waypoint destination using NED position vector from ekf origin in meters
-bool AC_WPNav::set_wp_destination_NED(const Vector3f& destination_NED)
-{
-    // convert NED to NEU and do not use terrain following
-    return set_wp_destination(Vector3f(destination_NED.x * 100.0f, destination_NED.y * 100.0f, -destination_NED.z * 100.0f), false);
+    set_kinematic_limits(_scurve_this_leg, _origin, _destination);
+    _scurve_this_leg.calculate_straight_leg(_origin, _destination);
+    _scurve_next_leg.init();
+    _flags.fast_waypoint = false;   // default waypoint back to slow
+    _flags.reached_destination = false;
+    _flags.wp_yaw_set = false;
+
+    return true;
 }
 
-/// set_origin_and_destination - set origin and destination waypoints using position vectors (distance from home in cm)
-///     terrain_alt should be true if origin.z and destination.z are desired altitudes above terrain (false if these are alt-above-ekf-origin)
-///     returns false on failure (likely caused by missing terrain data)
-bool AC_WPNav::set_wp_origin_and_destination(const Vector3f& origin, const Vector3f& destination, bool terrain_alt)
+/// set next destination using position vector (distance from ekf origin in cm)
+///     terrain_alt should be true if destination.z is a desired altitude above terrain
+///     provide next_destination
+bool AC_WPNav::set_wp_destination_next(const Vector3f& destination, bool terrain_alt)
 {
-    // store origin and destination locations
-    _origin = origin;
-    _destination = destination;
-    _terrain_alt = terrain_alt;
-    Vector3f pos_delta = _destination - _origin;
-
-    _track_length = pos_delta.length(); // get track length
-    _track_length_xy = safe_sqrt(sq(pos_delta.x)+sq(pos_delta.y));  // get horizontal track length (used to decide if we should update yaw)
-
-    // calculate each axis' percentage of the total distance to the destination
-    if (is_zero(_track_length)) {
-        // avoid possible divide by zero
-        _pos_delta_unit.x = 0;
-        _pos_delta_unit.y = 0;
-        _pos_delta_unit.z = 0;
-    }else{
-        _pos_delta_unit = pos_delta/_track_length;
+    // ToDo: improve handling of mismatching terrain alt
+    if (terrain_alt != _terrain_alt) {
+        return false;
     }
 
-    // calculate leash lengths
-    calculate_wp_leash_length();
+    set_kinematic_limits(_scurve_next_leg, _destination, destination);
+    _scurve_next_leg.calculate_straight_leg(_destination, destination);
 
-    // get origin's alt-above-terrain
-    float origin_terr_offset = 0.0f;
-    if (terrain_alt) {
-        if (!get_terrain_offset(origin_terr_offset)) {
-            return false;
-        }
-    }
+    // next destination provided so fast waypoint
+    _flags.fast_waypoint = true;
```

Comment:
> I can't help but think that scurves is really two separate classes forced to cohabit in one. All of the `switch` statements lead me to think this, and the fact that there exists `scurve.segtype` at all.
> 
> A different structure might have had a base class with two subclasses, one for straight, one for spline. WPNav would have pointers-to-instances of the base class and create a subclass object based on the next wp type (rather than calling `calculate_straight_leg` or `calculate_spline_leg` they'd create a subclass instance)
> 
> I'm not suggesting it be changed here. If the intent is that eventually these scurve objects could hold more than a single straight path or single spline then this entire comment can be disregarded.
> 
> I didn't go into the math in `calc_javp_` and friends, or in `cal_posfast` (and only lightly in many other places).

Yes I agree. When I wrote the straight line S-Curves it was really nice and clean. Splines messed it up. I suspect we will fix this when we fully understand what we want.

Reply:
It does seem to fly right to the waypoint.  I thought that it would fly onto to the edge of the radius around the waypoint but it seems to actually reach it.

![image](https://user-images.githubusercontent.com/1498098/108320172-c75b8980-7205-11eb-979d-bb79c7b06eef.png)